### ISSUE TITLE
[v5] AtomHashMap may produce invalid changes on Append

### ISSUE THREAD

**Issue by NonExisto** [louthy/language-ext#1412]
While studying project code base I've noticed occasional unit test failure for AtomHashMap. After spending some time I was able to replicate failure with following unit test theory:
```c#
    [Theory]
    [InlineData(-2)]
    [InlineData(0b10000011001)]
    [InlineData(0b1001100111001)]
    [InlineData(0b1001110000010)]
    [InlineData(4114)]
    public void AppendInvokesChangeInt(int conflictKey)
    {
        var                       hashMap      = AtomHashMap((conflictKey, 3), (2, 42));
        var                       toAppend     = HashMap((conflictKey, 6), (5, 7), (25, 9));
        var                       initialValue = hashMap.ToHashMap();
        HashMapPatch<int, int> state        = default!;
        hashMap.Change += v => state = v;

        hashMap.Append(toAppend);

        Assert.Equal(initialValue, state.From);
        Assert.Equal(hashMap.ToHashMap(), state.To);
        Assert.Equal(initialValue.Combine(toAppend), hashMap.ToHashMap());
        Assert.Equal(3, hashMap[conflictKey]);
        Assert.Equal(
            HashMap(
                (5, Change<int>.Added(7)),
                (25, Change<int>.Added(9))),
            state.Changes);
    }
```
Note: the state of `hashMap` is correct regardless of changes. It seems `ReferenceEquals(newRoot, Root)` inside of `UpdateWithLog` is not enough. Please have a look.

**Comment by louthy** [comment]
Ok, I'll take a look. I probably won't get to it until next week as I'm trying to make progress with my other projects

**Comment by louthy** [comment]
Fix in `v5.0.0-beta-39`
